distributed/tests/test_client.py

```diff
@@ -3592,65 +3592,56 @@ async def test_scatter_raises_if_no_workers(c, s):
-        async with Client(f"127.0.0.1:{port}", asynchronous=True) as c:
-            async with Worker(f"127.0.0.1:{port}") as w:
-                await c.wait_for_workers(1, timeout=10)
-                x = c.submit(inc, 1)
-                assert (await x) == 2
-                await hard_stop(s)
-
-                start = time()
-                while c.status != "connecting":
-                    assert time() < start + 10
-                    await asyncio.sleep(0.01)
-
-                assert x.status == "cancelled"
-                with pytest.raises(CancelledError):
-                    await x
-
-                async with Scheduler(port=port) as s2:
-                    start = time()
-                    while c.status != "running":
-                        await asyncio.sleep(0.1)
-                        assert time() < start + 10
-
-                    await w.finished()
-                    async with Worker(f"127.0.0.1:{port}"):
-                        start = time()
-                        while len(await c.nthreads()) != 1:
-                            await asyncio.sleep(0.05)
-                            assert time() < start + 10
-
-                        x = c.submit(inc, 1)
-                        assert (await x) == 2
-                        await hard_stop(s2)
-
-                start = time()
-                while True:
-                    assert time() < start + 10
-                    try:
-                        await x
-                        assert False
-                    except CommClosedError:
-                        continue
-                    except CancelledError:
-                        break
-            await c._close(fast=True)
+    stack = ExitStack()
+    proc = popen(["dask-scheduler", "--no-dashboard", f"--port={port}"])
+    stack.enter_context(proc)
+    async with Client(f"127.0.0.1:{port}", asynchronous=True) as c, Worker(
+        f"127.0.0.1:{port}"
+    ) as w:
+        await c.wait_for_workers(1, timeout=10)
+        x = c.submit(inc, 1)
+        assert (await x) == 2
+        stack.close()
+
+        start = time()
+        while c.status != "connecting":
+            assert time() < start + 10
+            await asyncio.sleep(0.01)
+
+        assert x.status == "cancelled"
+        with pytest.raises(CancelledError):
+            await x
+
+        stack2 = ExitStack()
+        proc2 = popen(["dask-scheduler", "--no-dashboard", f"--port={port}"])
+        stack2.enter_context(proc2)
+        start = time()
+        while c.status != "running":
+            await asyncio.sleep(0.1)
+            assert time() < start + 10
+
+        await w.finished()
+        async with Worker(f"127.0.0.1:{port}"):
+            start = time()
+            while len(await c.nthreads()) != 1:
+                await asyncio.sleep(0.05)
+                assert time() < start + 10
+
+            x = c.submit(inc, 1)
+            assert (await x) == 2
+            stack2.close()
```

Comment:
```suggestion
        with popen(["dask-scheduler", "--no-dashboard", f"--port={port}"]):
            start = time()
            while c.status != "running":
                await asyncio.sleep(0.1)
                assert time() < start + 10

            await w.finished()
            async with Worker(f"127.0.0.1:{port}"):
                start = time()
                while len(await c.nthreads()) != 1:
                    await asyncio.sleep(0.05)
                    assert time() < start + 10

                x = c.submit(inc, 1)
                assert (await x) == 2
```
Not sure you need the second exit stack?

Reply:
don't need it. copied it